Copy limits rework

rafatower opened the pull request:
This does a number of things:
- It sets the COPY FROM/TO to use the Batch API port, instead of the regular port (that is, use postgres directly instead of going through pgbouncer)
- It sets a specific configurable timeout for the COPY queries, set by default to 5h.
- It adds a middleware to check for the remaining DB quota, and it makes the COPY FROM fail in case it gets a bigger payload than the remaining quota.
- It adds a hard limit on the max size of the payload received by the COPY FROM. Actually that is a limit imposed by nginx configuration, but we want to catch and report a proper error in case we're close to the limit.

What is remaining: test in staging ~~and update the documentation (already being used) with the limits~~ (update: docs modified to reflect new limits).

Requesting a review cause I'm pretty sure you'll have suggestions for improvement and would like to know about them sooner rather than later :slightly_smiling_face:

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -101,8 +100,10 @@ function handleCopyTo (logger) {
-        const { userDbParams, user } = res.locals;
+        const { userDbParams, user, dbRemainingQuota } = res.locals;
         const isGzip = req.get('content-encoding') === 'gzip';
+        const copy_from_max_post_size = global.settings.copy_from_max_post_size || 1.99 * 1024 * 1024 * 1024; // 1.99 GB
```
It would be nice using capital letters.

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -101,8 +100,10 @@ function handleCopyTo (logger) {
+        const { userDbParams, user, dbRemainingQuota } = res.locals;
         const isGzip = req.get('content-encoding') === 'gzip';
+        const copy_from_max_post_size = global.settings.copy_from_max_post_size || 1.99 * 1024 * 1024 * 1024; // 1.99 GB
+        const copy_from_max_post_size_pretty = global.settings.copy_from_max_post_size_pretty || '2 GB';
```
Same: capital letters

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -124,7 +125,24 @@ function handleCopyFrom (logger) {
+                    metrics.addSize(data.length);
+
+                    if(metrics.size > dbRemainingQuota) {
+                        const quotaError = new Error('DB Quota exceeded');
```
In order to handle the error in 1 place, we used the following way:
```
const quotaError = new Error('DB Quota exceeded');
pgstream.emit('error', quotaError);
```

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -124,7 +125,24 @@ function handleCopyFrom (logger) {
+                        return next(quotaError);
+                    }
+                    if((metrics.gzipSize || metrics.size) > copy_from_max_post_size) {
+                        const maxPostSizeError = new Error(
```
Same as before:
```
const maxPostSizeError = new Error(...
pgstream.emit('error', maxPostSizeError);
```

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -124,7 +125,24 @@ function handleCopyFrom (logger) {
+                        req.unpipe(pgstream);
+                        return next(quotaError);
+                    }
+                    if((metrics.gzipSize || metrics.size) > copy_from_max_post_size) {
```
Elegant :muscle:

oleurud commented on app/middlewares/db-quota.js:
```diff
@@ -0,0 +1,24 @@
+const PSQL = require('cartodb-psql');
+
+const remainingQuotaQuery = 'SELECT _CDB_UserQuotaInBytes() - CDB_UserDataSize(current_schema()) AS remaining_quota';
```
:muscle:

oleurud commented on app/middlewares/db-quota.js:
```diff
@@ -0,0 +1,24 @@
+                if(err) {
+                    return next(err);
+                }
+                const remainingQuota = result.rows[0].remaining_quota;
```
Only a comment, not sure if it's a problem: do we always have a number here? does the query always return a value? Does the user always have a quota?

We talked about it in the past and I finished believing it does not always available

rafatower replied:
Yes, there is always a quota set.

It is the same query as launched by Builder and Editor to check the user quota, but done in a simpler way that does not require extra metadata:
- regular users have their `cartodb-postgresql` stuff in the `public` schema.
- organization users have their `cartodb-postgresql` stuff in the `username` schema.
- both of them have their main schema (`public` or `username`) in the front of their `search_path`.

I checked it and it works for regular and organizations users as expected.

enekid commented on app/services/stream_copy.js:
```diff
@@ -27,27 +33,34 @@ module.exports = class StreamCopy {
+                            cancelingClient.cancel(client, pgstream);
+
+                            // see https://node-postgres.com/api/pool#releasecallback
+                            done(err);
```
Just to prevent future errors, I would add a `return` here:
```javascript
return done(err);
```

rafatower replied:
good catch

enekid commented on app/services/stream_copy.js:
```diff
@@ -27,27 +33,34 @@ module.exports = class StreamCopy {
+                    return cb(err);
+                }
+
+                let streamMaker = action === ACTION_TO ? copyTo : copyFrom;
```
`const`?
```javascript
const streamMaker = action === ACTION_TO ? copyTo : copyFrom;
```

rafatower replied:
PI and E are constants, I cannot tell about other things... Ooookay :stuck_out_tongue:

enekid replied:
It's JavaScript. `const` is not really `const`. It's something like a `const`  😮

rafatower commented:
@enekid et al: I addressed your comments so far, in case you want to take a second look at it

rafatower commented:
@dgaubert could my code have your blessing? I plan to test it in staging this afternoon, time permitting.

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -124,7 +125,22 @@ function handleCopyFrom (logger) {
+
+                    if(metrics.size > dbRemainingQuota) {
+                        const quotaError = new Error('DB Quota exceeded');
+                        metrics.end(null, quotaError);
```
It won't be necessary (it is made in https://github.com/CartoDB/CartoDB-SQL-API/pull/523/files#diff-373cc3192301f2293b950b57676c109dR147)

oleurud commented on app/controllers/copy_controller.js:
```diff
@@ -124,7 +125,22 @@ function handleCopyFrom (logger) {
+                        const maxPostSizeError = new Error(
+                            `COPY FROM maximum POST size of ${COPY_FROM_MAX_POST_SIZE_PRETTY} exceeded`
+                        );
+                        metrics.end(null, maxPostSizeError);
```
Same, not necessary

rafatower commented:
### Tests in staging

Branch deployed to https://rtorre-copy-limits.carto-staging.com with a couple org users.

Test plan (basically reproduce the automated tests, with a few changes to config):
- [ ] `./copyfrom` sunny day scenario
- [ ] `/copyto` sunny day scenario
- [ ] Reduce the test user's db quota and check that the "quota exceeded" error is triggered.
- [ ] Set a small value for `copy_from_max_post_size` and check that the max post size error is triggered.
- [ ] Check again that with those reduced quota and max_post_size values the `./copyto` is not affected.

rafatower commented:
### Test logs

I used the scripts in https://github.com/CartoDB/cartodb-tests/tree/master/sql_copy_perf

Due to a varnish misconfiguration issue in staging, I had to patch the scripts to go through `api/v1/sql`.


<details>

* `/copyfrom` sunny day scenario:

With the user `copy-limits-1` with API KEY `e0bdc3ac0a5cf92702a85c0defb216e2ba0483dd`

```
$ ./points.sh https://copy-limits-1.carto-staging.com e0bdc3ac0a5cf92702a85c0defb216e2ba0483dd 100
...
{"time":0.066,"total_rows":100}
```

* `./copyto` sunny day scenario:

Same user.

```
* ./export-table.sh https://copy-limits-1.carto-staging.com e0bdc3ac0a5cf92702a85c0defb216e2ba0483dd test_copy_points -o export_test_copy_points2.csv -f csv
...
{ [5411 bytes data]
100  5392    0  5392    0     0   7445      0 --:--:-- --:--:-- --:--:--  7437
* Connection #0 to host copy-limits-1.carto-staging.com left intact

    time_namelookup:  0.253
       time_connect:  0.301
    time_appconnect:  0.549
   time_pretransfer:  0.549
      time_redirect:  0.000
 time_starttransfer:  0.724
                    ----------
         time_total:  0.724
```

* Reduce the test user's db quota and check that the "quota exceeded" error is triggered.


I updated the user's quota in superadmin: `24.0 of 25.0 used KB`

Then I ran the very first script again, with 1K rows:

```
$ ./points.sh https://copy-limits-1.carto-staging.com e0bdc3ac0a5cf92702a85c0defb216e2ba0483dd 1000
...
< HTTP/1.1 100 Continue
< HTTP/1.1 400 Bad Request
< Server: openresty
< Date: Tue, 24 Jul 2018 13:37:08 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 31
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: X-Requested-With, X-Prototype-Version, X-CSRF-Token
< X-Served-By-Host: ded-dynamic-p2KCS4
< vary: Authorization
< Content-Disposition: inline
< X-SQLAPI-Profiler: {"authorization":2,"getConnectionParams":4,"finish":52,"total":58}
< X-SQLAPI-Errors: {"statusCode":400,"message":"DB Quota exceeded"}
< 
* Connection #0 to host copy-limits-1.carto-staging.com left intact
{"error":["DB Quota exceeded"]}
```

* Set a small value for copy_from_max_post_size and check that the max post size error is triggered:

I modified the config in the staging, adding this:

```js
// file www/node-sql-api/shared/config/staging.js
module.exports.copy_from_max_post_size = 10;
module.exports.copy_from_max_post_size_pretty = '10 ridiculous bytes';
```

and ran the very first script again:

```
$ ./points.sh https://copy-limits-1.carto-staging.com e0bdc3ac0a5cf92702a85c0defb216e2ba0483dd 100

< HTTP/1.1 100 Continue
< HTTP/1.1 400 Bad Request
< Server: openresty
< Date: Tue, 24 Jul 2018 13:46:47 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 73
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: X-Requested-With, X-Prototype-Version, X-CSRF-Token
< X-Served-By-Host: ded-dynamic-p2KCS4
< vary: Authorization
< Content-Disposition: inline
< X-SQLAPI-Profiler: {"authorization":1,"getConnectionParams":1,"finish":22,"total":24}
< X-SQLAPI-Errors: {"statusCode":400,"message":"COPY FROM maximum POST size of 10 ridiculous bytes exceeded"}
< 
* Connection #0 to host copy-limits-1.carto-staging.com left intact
{"error":["COPY FROM maximum POST size of 10 ridiculous bytes exceeded"]}
```

* Check again that with those reduced quota and `max_post_size` values, the `./copyto` is not affected:

(I had to load the data directly into the staging db, cause the original test scrips truncated everything, in order to avoid resetting limits)

```
$ ./export-table.sh https://copy-limits-1.carto-staging.com e0bdc3ac0a5cf92702a85c0defb216e2ba0483dd test_copy_points -o export_test_copy_points2.csv -f csv
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0< HTTP/1.1 200 OK
< Server: openresty
< Date: Tue, 24 Jul 2018 13:58:37 GMT
< Content-Type: application/octet-stream
< Transfer-Encoding: chunked
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: X-Requested-With, X-Prototype-Version, X-CSRF-Token
< X-Served-By-Host: ded-dynamic-p2KCS4
< vary: Authorization
< Content-Disposition: attachment; filename=carto-sql-copyto.dmp
< 
{ [15979 bytes data]
100 53390    0 53390    0     0  60723      0 --:--:-- --:--:-- --:--:-- 60670
* Connection #0 to host copy-limits-1.carto-staging.com left intact

    time_namelookup:  0.254
       time_connect:  0.300
    time_appconnect:  0.566
   time_pretransfer:  0.566
      time_redirect:  0.000
 time_starttransfer:  0.830
                    ----------
         time_total:  0.879

```

</details>

rafatower commented:
And a couple extra, missing tests:
- [ ] check that the regular timeout for queries does not bother with a slow connection
- [ ] check that the timeout for copy can be configured, and it kicks in when needed.
